Development/nmos/client_utils.cpp

```diff
@@ -55,39 +60,165 @@ namespace nmos
+        {
+            int socket_fd;
+            if (secure)
+            {
+                auto socket = (boost::asio::ssl::stream<boost::asio::ip::tcp::socket&>*)native_handle;
+                if (!socket->lowest_layer().is_open())
+                {
+                    // for now, limited to IPv4
+                    socket->lowest_layer().open(boost::asio::ip::tcp::v4());
+                }
+                socket_fd = socket->lowest_layer().native_handle();
+            }
+            else
+            {
+                auto socket = (boost::asio::ip::tcp::socket*)native_handle;
+                if (!socket->is_open())
+                {
+                    // for now, limited to IPv4
+                    socket->open(boost::asio::ip::tcp::v4());
+                }
+                socket_fd = socket->lowest_layer().native_handle();
+            }
+            // SO_BINDTODEVICE not defined in windows & mac
+            return (setsockopt(socket_fd, SOL_SOCKET, SO_BINDTODEVICE, utility::us2s(interface_name).c_str(), interface_name.length()) == 0);
+        }
+#endif
+
+        inline std::function<void(web::http::client::native_handle)> make_client_nativehandle_options(const utility::string_t& client_address, bool secure, slog::base_gate& gate)
+        {
+            // get the associated network interface name from IP address
+            const auto interface_name = get_interface_name(client_address);
+            if (!client_address.empty() && interface_name.empty())
+            {
+                slog::log<slog::severities::error>(gate, SLOG_FLF) << "No network interface found for " << client_address << " to bind for the HTTP client connection";
+            }
+
+            return [interface_name, secure, &gate](web::http::client::native_handle native_handle)
+            {
+                if (!interface_name.empty())
+                {
+#if defined(linux) || defined(__linux) || defined(__linux__)
+                    if (!bind_to_device(interface_name, secure, native_handle))
+                    {
+                        char error[1024];
+                        slog::log<slog::severities::error>(gate, SLOG_FLF) << "Unable to bind HTTP client connection to " << interface_name << ", bind_to_device reported error : " << strerror_r(errno, error, sizeof(error));
+                    }
+#else
+                    slog::log<slog::severities::error>(gate, SLOG_FLF) << "Bind HTTP client connection to " << interface_name << " not supported";
+#endif
+                }
+            };
+        }
+
+#ifdef CPPRESTSDK_ENABLE_BIND_WEBSOCKET_CLIENT
+        // The current version of the C++ REST SDK 2.10.18 does not provide the callback to enable the custom websocket setting
+        inline std::function<void(web::websockets::client::native_handle)> make_ws_client_nativehandle_options(const utility::string_t& client_address, bool secure, slog::base_gate& gate)
```

Comment:
Will `web::websockets::client::native_handle` be same underlying type as `web::http::client::native_handle`? I suppose not necessarily...
Just seems a lot of duplication between these functions.

Reply:
From my understanding, they are the same at least for the Linux implementation. see the diff of cpprestsdk. 

```
 Release/include/cpprest/ws_client.h              | 26 ++++++++++++++++++++++++
 Release/src/websockets/client/ws_client_wspp.cpp | 12 +++++++++++
 2 files changed, 38 insertions(+)

diff --git a/Release/include/cpprest/ws_client.h b/Release/include/cpprest/ws_client.h
index af17bd60..74d8c9c0 100644
--- a/Release/include/cpprest/ws_client.h
+++ b/Release/include/cpprest/ws_client.h
@@ -69,6 +69,8 @@ enum class websocket_close_status
     server_terminate = 1011,
 };
 
+typedef void* native_handle;
+
 /// <summary>
 /// Websocket client configuration class, used to set the possible configuration options
 /// used to create an websocket_client instance.
@@ -179,6 +181,29 @@ public:
     /// caution.</remarks>
     void set_validate_certificates(bool validate_certs) { m_validate_certificates = validate_certs; }
 
+    /// <summary>
+    /// Sets a callback to enable custom setting of platform specific options.
+    /// </summary>
+    /// <remarks>
+    /// The native_handle is the following type depending on the underlying platform:
+    ///         https - boost::asio::ssl::stream<boost::asio::ip::tcp::socket &> *
+    ///         http - boost::asio::ip::tcp::socket *
+    /// </remarks>
+    /// <param name="callback">A user callback allowing for customization of the connect</param>
+    void set_nativehandle_options(const std::function<void(native_handle)>& callback)
+    {
+        m_set_user_nativehandle_options = callback;
+    }
+
+    /// <summary>
+    /// Invokes a user's callback to allow for customization of the connect.
+    /// </summary>
+    /// <param name="handle">A internal implementation handle.</param>
+    void invoke_nativehandle_options(native_handle handle) const
+    {
+        if (m_set_user_nativehandle_options) m_set_user_nativehandle_options(handle);
+    }
+
 #if !defined(_WIN32) || !defined(__cplusplus_winrt)
     /// <summary>
     /// Sets a callback to enable custom setting of the ssl context, at construction time.
@@ -209,6 +234,7 @@ private:
 #if !defined(_WIN32) || !defined(__cplusplus_winrt)
     std::function<void(boost::asio::ssl::context&)> m_ssl_context_callback;
 #endif
+    std::function<void(native_handle)> m_set_user_nativehandle_options;
 };
 
 /// <summary>
diff --git a/Release/src/websockets/client/ws_client_wspp.cpp b/Release/src/websockets/client/ws_client_wspp.cpp
index d7c31c40..825ae465 100644
--- a/Release/src/websockets/client/ws_client_wspp.cpp
+++ b/Release/src/websockets/client/ws_client_wspp.cpp
@@ -262,6 +262,8 @@ public:
                         SSL_set_tlsext_host_name(ssl_stream.native_handle(), const_cast<char*>(server_name.c_str()));
                     }
                 }
+
+                m_config.invoke_nativehandle_options(&ssl_stream);
             });
 
             return connect_impl<websocketpp::config::asio_tls_client>();
@@ -269,6 +271,16 @@ public:
         else
         {
             m_client = std::unique_ptr<websocketpp_client_base>(new websocketpp_client());
+
+            // Options specific to underlying socket.
+            auto& client = m_client->client<websocketpp::config::asio_client>();
+
+            client.set_socket_init_handler([this](websocketpp::connection_hdl,
+                                                  boost::asio::ip::tcp::socket& socket) {
+
+                m_config.invoke_nativehandle_options(&socket);
+            });
+
             return connect_impl<websocketpp::config::asio_client>();
         }
```
     }